File: typescript/packages/core/src/core/github/release.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { GitHubApiError, type GitHubTransport } from './client.ts'
import { githubPages } from './paginate.ts'
import type { RepoRef } from './repo.ts'

function path(ref: RepoRef, tail = ''): string {
  return `/repos/${ref.owner}/${ref.repo}/releases${tail}`
}

export function listReleases(
  transport: GitHubTransport,
  ref: RepoRef,
  limit: number,
): Promise<Record<string, unknown>[]> {
  return githubPages(transport, path(ref), { limit })
}

export function getRelease(
  transport: GitHubTransport,
  ref: RepoRef,
  tag: string,
): Promise<unknown> {
  return transport.get(path(ref, `/tags/${encodeURIComponent(tag)}`))
}

export async function getLatestRelease(transport: GitHubTransport, ref: RepoRef): Promise<unknown> {
  try {
    return await transport.get(path(ref, '/latest'))
  } catch (err) {
    if (err instanceof GitHubApiError && err.status === 404) return null
    throw err
  }
}

export function createRelease(
  transport: GitHubTransport,
  ref: RepoRef,
  body: Record<string, unknown>,
): Promise<unknown> {
  return transport.request('POST', path(ref), body)
}

File: python/mirage/core/github/release.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from typing import Any
from urllib.parse import quote

from mirage.core.github.client import GitHubApiError, github_request
from mirage.core.github.config import GhConfig
from mirage.core.github.paginate import github_pages
from mirage.core.github.repo import RepoRef
from mirage.types import JsonValue


def _path(ref: RepoRef, tail: str = "") -> str:
    return f"/repos/{ref.owner}/{ref.repo}/releases{tail}"


async def list_releases(config: GhConfig, ref: RepoRef,
                        limit: int) -> list[dict[str, Any]]:
    return await github_pages(config, _path(ref), limit=limit)


async def get_release(config: GhConfig, ref: RepoRef, tag: str) -> JsonValue:
    return await github_request(config.token,
                                "GET",
                                _path(ref, f"/tags/{quote(tag, safe='')}"),
                                base_url=config.base_url)


async def get_latest_release(config: GhConfig,
                             ref: RepoRef) -> "JsonValue | None":
    try:
        return await github_request(config.token,
                                    "GET",
                                    _path(ref, "/latest"),
                                    base_url=config.base_url)
    except GitHubApiError as exc:
        if exc.status == 404:
            return None
        raise


async def create_release(config: GhConfig, ref: RepoRef,
                         body: dict[str, JsonValue]) -> JsonValue:
    return await github_request(config.token,
                                "POST",
                                _path(ref),
                                body,
                                base_url=config.base_url)
